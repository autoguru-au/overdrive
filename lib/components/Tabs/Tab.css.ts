import { style } from '@vanilla-extract/css';
import { recipe } from '@vanilla-extract/recipes';

import { overdriveTokens as vars } from '../../themes/theme.css';

import { focusOutline } from './../../styles/focusOutline.css';

const lineBottomHeight = '1px';
const size = '20px';

export const styledTab = recipe({
	base: {
		flex: 'auto',
		transition: `color 0.2s ${vars.animation.easing.decelerate} 0s, background-color 0.2s ${vars.animation.easing.decelerate} 0s`,
		':focus-visible': {
			...focusOutline,
		},
		selectors: {
			'&:not([aria-selected=true]):hover': {
				color: vars.colours.foreground.body,
			},
		},
	},
	variants: {
		appearance: {
			underlined: {
				borderBottom: `calc(${lineBottomHeight} + ${lineBottomHeight}) solid transparent`,
				padding: `calc(${vars.space['3']} + ${lineBottomHeight}) ${vars.space['4']}`,
				':focus-visible': {
					outlineOffset: '-1px',
				},
			},
			pill: {
				borderRadius: vars.border.radius.pill,
				color: vars.colours.foreground.body,
				fontWeight: vars.typography.fontWeight.normal,
				padding: `${vars.space['2']} ${vars.space['4']}`,
				selectors: {
					'&+&': {
						marginLeft: vars.space['3'],
					},
					'&:not([aria-selected=true]):hover': {
						backgroundColor: vars.colours.gamut.gray200,
					},
				},
			},
		},
		active: {
			true: {},
		},
	},
	compoundVariants: [
		{
			variants: {
				appearance: 'underlined',
				active: true,
			},
			style: {
				color: vars.colours.intent.neutral.background.strong,
				borderBottomColor:
					vars.colours.intent.neutral.background.strong,
			},
		},
		{
			variants: {
				appearance: 'pill',
				active: true,
			},
			style: {
				backgroundColor: vars.colours.foreground.body,
				color: vars.colours.background.body,
			},
		},
	],
	defaultVariants: {
		appearance: 'underlined',
	},
});

export const item = style({
	display: 'inline-flex',
	verticalAlign: 'middle',
	width: 'max-content',
});

export const indication = recipe({
	base: [
		{
			backgroundColor: vars.colours.background.light,
			borderRadius: vars.border.radius.pill,
			display: 'inline-block',
			minWidth: size,
			height: size,
			padding: `0 ${vars.space['1']}`,
			lineHeight: size,
			transition: `color 0.2s ${vars.animation.easing.decelerate} 0s, backgroundColor 0.2s ${vars.animation.easing.decelerate} 0s`,
		},
	],
	variants: {
		appearance: {
			underlined: {},
			pill: {
				selectors: {
					[`${styledTab.classNames.base}:not([aria-selected=true]):hover &`]:
						{
							backgroundColor: vars.colours.background.neutral,
						},
				},
			},
		},
		active: {
			true: {},
		},
	},
	compoundVariants: [
		{
			variants: {
				appearance: 'underlined',
				active: true,
			},
			style: {
				backgroundColor: vars.colours.intent.neutral.background.strong,
			},
		},
		{
			variants: {
				appearance: 'pill',
				active: true,
			},
			style: {
				backgroundColor: vars.colours.background.light,
				color: vars.colours.foreground.body,
			},
		},
	],
	defaultVariants: {
		appearance: 'underlined',
	},
});
