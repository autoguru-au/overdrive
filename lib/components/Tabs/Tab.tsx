import { FunctionComponent } from 'react';

export interface Props {
	title: string;
	indication?: number;
}

export const Tab: FunctionComponent<Props> = () => {
	throw new Error("The Tab component isn't mean to be rendered by itself...");
};
