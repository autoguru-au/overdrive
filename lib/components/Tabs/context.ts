import { createContext } from 'react';

export const TabsContext = createContext<{
	id: string | null;
	active: number;
	onChange?: (idx: number) => void;
}>({
	active: 0,
	id: null,
});

export const IndexContext = createContext<number | null>(null);
