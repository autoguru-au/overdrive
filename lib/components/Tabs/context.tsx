import * as React from 'react';
import {
	ContextType,
	createContext,
	FunctionComponent,
	useContext,
} from 'react';

const TabsContext = createContext<{
	id: string | null;
	active: number;
	onChange?: (idx: number) => void;
}>({
	active: 0,
	id: null,
});

export const useTabsContext = () => useContext(TabsContext);
export const TabsContextProvider: FunctionComponent<ContextType<
	typeof TabsContext
>> = ({ active, id, onChange, children }) => (
	<TabsContext.Provider
		value={{
			active,
			id,
			onChange,
		}}>
		{children}
	</TabsContext.Provider>
);

const IndexContext = createContext<number | null>(null);

export const TabIndexProvider: FunctionComponent<{ index: number }> = ({
	index,
	children,
}) => <IndexContext.Provider value={index}>{children}</IndexContext.Provider>;
export const useTabIndex = () => useContext(IndexContext);
