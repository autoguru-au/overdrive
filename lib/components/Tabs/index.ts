export { Tabs } from './Tabs';
export { Tab } from './Tab';
export { TabList } from './TabList';
export { TabPane } from './TabPane';
export { TabPanes } from './TabPanes';
