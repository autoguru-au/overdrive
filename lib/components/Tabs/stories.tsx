import * as React from 'react';
import { useEffect, useState } from 'react';
import isChromatic from 'storybook-chromatic/isChromatic';

import { Stack } from '../Stack';
import { StarRating } from '../StarRating';
import { Tab, TabList, TabPane, TabPanes, Tabs } from '.';

export default { title: 'Components|Tabs' };

export const Standard = () => (
	<Tabs>
		<TabList>
			<Tab>Tab 1</Tab>
			<Tab>Tab 2</Tab>
		</TabList>

		<TabPanes>
			<TabPane>Content A</TabPane>
			<TabPane>
				<Stack>
					<TestChild label="5" />
					<TestChild label="4" />
					<TestChild label="3" />
					<TestChild label="2" />
					<TestChild label="1" />
				</Stack>
			</TabPane>
		</TabPanes>
	</Tabs>
);

export const WithIndication = () => (
	<Tabs>
		<TabList>
			<Tab indication={2}>Tab 1</Tab>
			<Tab indication={0}>Tab 2</Tab>
		</TabList>

		<TabPanes>
			<TabPane>Content A</TabPane>
			<TabPane>Content B</TabPane>
		</TabPanes>
	</Tabs>
);

export const TabsWithoutPanes = () => (
	<>
		<Tabs>
			<TabList>
				<Tab id="tab-1" is={<a href="#tab-1" />}>
					Tab 1
				</Tab>
				<Tab id="tab-2" is={<a href="#tab-2" />}>
					Tab 2
				</Tab>
			</TabList>
		</Tabs>

		<ul>
			<li id="tab-1">Tab 1 Content</li>
			<li id="tab-2" style={{ marginTop: '120vh' }}>
				Tab 2 Content
			</li>
		</ul>
	</>
);

const TestChild = ({ label }) => {
	const [thing, sething] = useState(isChromatic() ? 0.5 : Math.random() * 5);

	useEffect(() => {
		const t = setInterval(() => {
			sething(isChromatic() ? 0.5 : Math.random() * 5);
		}, 1e3);

		return () => {
			clearInterval(t);
		};
	}, []);

	return <StarRating rating={thing} label={label} />;
};
