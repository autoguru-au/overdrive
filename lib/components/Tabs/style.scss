@import '../../theme/tokens/facade';

$lineBottomHeight: 2px;

.nav-item {
	$sel: &;

	display: block;
	outline: none;
	width: auto;
	border: none;
	margin: 0 0 (-1 * $lineBottomHeight);
	background: transparent;
	font-size: var(--global--fontSize--4);
	font-weight: 700;
	line-height: 18px;
	padding: calc(var(--global--space--3) + 1px) 0;
	color: getColour(gray, 500);
	cursor: pointer;
	border-bottom: $lineBottomHeight solid transparent;
	transition: color 0.2s cubic-bezier(0, 0, 0.2, 1) 0s,
		background-color 0.2s cubic-bezier(0, 0, 0.2, 1) 0s;

	&:not(:last-of-type) {
		margin-right: var(--global--space--6);
	}

	&:hover,
	&:focus {
		color: getColour(green, 900);
	}

	&--active {
		color: getColour(green, 900);
		border-bottom-color: getColour(green, 900);
	}

	&__title {
		display: inline-block;
		vertical-align: middle;
	}

	&__indication {
		display: inline-block;
		vertical-align: middle;
		border-radius: 50%;
		width: 18px;
		height: 18px;
		font-size: calc(var(--global--fontSize--4) * 0.8);
		color: getColour(gray, 900);
		background-color: getColour(gray, 300);
		margin-left: var(--global--space--2);
		transition: color 0.2s cubic-bezier(0, 0, 0.2, 1) 0s,
			background-color 0.2s cubic-bezier(0, 0, 0.2, 1) 0s;

		@at-root #{selector_nest($sel + '--active', &),
			#{selector_nest($sel + ':hover', &)}} {
			color: white;
			background-color: getColour(green, 900);
		}
	}
}

.tab-pane {
	display: none;
	float: left;
	width: 100%;

	&--active {
		display: block;
	}
}

.tabs {
	display: block;
	width: 100%;

	&__nav {
		display: flex;
		flex-wrap: nowrap;
		border-bottom: $lineBottomHeight solid getColour(gray, 300);
	}

	&__content {
		padding: var(--global--space--6) 0;

		&:after {
			content: '';
			display: table;
			clear: both;
		}
	}
}
