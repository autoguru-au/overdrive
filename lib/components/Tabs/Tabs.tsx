import React, {
	Children,
	ComponentPropsWithoutRef,
	FunctionComponent,
	ReactElement,
	useEffect,
	useState,
} from 'react';

import styles from './style.scss';
import { Tab } from './Tab';
import { TabNavItem } from './TabNavItem';
import { TabPane } from './TabPane';

export interface Props {
	active?: number;

	onChange?(value: number): void;
}

export const Tabs: FunctionComponent<Props> = ({
	active: incomingActive = 0,
	onChange = () => void 0,
	children,
}) => {
	const [active, setActive] = useState(incomingActive);
	const [prevActive, setPrevActive] = useState(null);

	if (incomingActive !== prevActive) {
		setActive(incomingActive);
		setPrevActive(incomingActive);
	}

	const setActiveCb = idx => () => {
		onChange(idx);
		setActive(idx);
	};

	const tabsChildren = Children.toArray(children) as Array<
		ReactElement<ComponentPropsWithoutRef<typeof Tab>>
	>;

	const tabId = useId();

	const tabs = tabsChildren.map((child, idx) => {
		const tabItemId = `tabItem:${tabId}:${idx}`;
		const tabPanelId = `tabPanel:${tabId}:${idx}`;

		return [
			<TabNavItem
				key={idx}
				id={tabItemId}
				active={active === idx}
				indication={child.props.indication}
				onClick={setActiveCb(idx)}>
				{child.props.title}
			</TabNavItem>,
			<TabPane
				key={idx}
				id={tabPanelId}
				controlledBy={tabItemId}
				active={active === idx}>
				{child.props.children}
			</TabPane>,
		];
	});

	return (
		<div className={styles.tabs} role="tablist">
			<div className={styles.tabsNav}>{tabs.map(([item]) => item)}</div>
			<div className={styles.tabsContent}>
				{tabs.map(([, item]) => item)}
			</div>
		</div>
	);
};

let id = 0;
const genId = () => ++id;

const useId = () => {
	const [id, setId] = useState(null);
	useEffect(() => setId(genId()), []);
	return `od-${id}`;
};
