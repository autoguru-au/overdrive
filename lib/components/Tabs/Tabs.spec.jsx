import React from 'react';
import { mount, shallow } from 'enzyme';
import { Tabs } from './Tabs';
import { act } from 'react-dom/test-utils';
import { Tab } from './Tab';
import ReactDOM from 'react-dom';

describe('<Tabs />', () => {
	const tabData = [
		{
			title: 'tab 1 title',
			content: 'tab 1 content',
		},
		{
			title: 'tab 2 title',
			content: 'tab 2 content',
		},
		{
			title: 'tab 3 title',
			content: 'tab 3 content',
		},
	];

	it('should throw if a tab element is not nested inside a Tabs element', () =>
		expect(() =>
			shallow(
				<Tab title="tab">
					<p>Tab content</p>
				</Tab>,
			),
		).toThrow());

	it('should not throw', () => expect(() => shallow(<Tabs />)).not.toThrow());

	it('should not throw with nested tab items', () =>
		expect(() =>
			shallow(
				<Tabs>
					{tabData.map(tabData => (
						<Tab key={tabData} title={tabData.title}>
							{tabData.content}
						</Tab>
					))}
				</Tabs>,
			),
		).not.toThrow());

	it('should match snapshot with nested tab items', () => {
		const tabs = mount(
			<Tabs>
				{tabData.map(tabData => (
					<Tab key={tabData} title={tabData.title}>
						{tabData.content}
					</Tab>
				))}
			</Tabs>,
		);
		expect(tabs.html()).toMatchSnapshot();

		tabs.unmount();
	});

	it('should display the first tab pane by default', () => {
		const tabs = mount(
			<Tabs>
				{tabData.map(tabData => (
					<Tab key={tabData} title={tabData.title}>
						{tabData.content}
					</Tab>
				))}
			</Tabs>,
		);

		expect(
			tabs
				.find('.tabPane')
				.at(0)
				.hasClass('tabPaneActive'),
		).toEqual(true);
		expect(
			tabs
				.find('.tabPane')
				.at(1)
				.hasClass('tabPaneActive'),
		).toEqual(false);
		expect(
			tabs
				.find('.tabPane')
				.at(2)
				.hasClass('tabPaneActive'),
		).toEqual(false);

		tabs.unmount();
	});

	it('should allow the active to be updated outside', () => {
		const tabs = mount(
			<Tabs>
				<Tab title="A">A</Tab>
				<Tab title="B">B</Tab>
			</Tabs>,
		);

		expect(
			tabs
				.find('.tabPane')
				.at(0)
				.hasClass('tabPaneActive'),
		).toEqual(true); // Precondition

		tabs.setProps({ active: 1 });

		expect(
			tabs
				.find('.tabPane')
				.at(1)
				.hasClass('tabPaneActive'),
		).toEqual(true);
	});

	describe('when rendered to the DOM', () => {
		let container;

		const evt = new Event('click', {
			bubbles: true,
			cancelable: false,
			composed: false,
		});

		beforeEach(function() {
			container = document.createElement('div');
			document.body.append(container);
		});

		afterEach(function() {
			document.body.removeChild(container);
		});

		it('should only display the first tab pane when tab 1 is clicked', done => {
			act(() => {
				ReactDOM.render(
					<Tabs>
						{tabData.map(tabData => (
							<Tab key={tabData} title={tabData.title}>
								{tabData.content}
							</Tab>
						))}
					</Tabs>,
					container,
				);
			});

			act(() => {
				document
					.querySelector('.tabs button:nth-child(1)')
					.dispatchEvent(evt);

				setTimeout(() => {
					expect(
						document
							.querySelector('.tabPane:nth-child(1)')
							.className.includes('tabPaneActive'),
					).toEqual(true);
					expect(
						document
							.querySelector('.tabPane:nth-child(2)')
							.className.includes('tabPaneActive'),
					).toEqual(false);
					expect(
						document
							.querySelector('.tabPane:nth-child(3)')
							.className.includes('tabPaneActive'),
					).toEqual(false);
					done();
				});
			});
		});

		it('should only display the second tab pane when tab 2 is clicked', done => {
			act(() => {
				ReactDOM.render(
					<Tabs>
						{tabData.map(tabData => (
							<Tab key={tabData} title={tabData.title}>
								{tabData.content}
							</Tab>
						))}
					</Tabs>,
					container,
				);
			});

			act(() => {
				document
					.querySelector('.tabs button:nth-child(2)')
					.dispatchEvent(evt);

				setTimeout(() => {
					expect(
						document
							.querySelector('.tabPane:nth-child(1)')
							.className.includes('tabPaneActive'),
					).toEqual(false);
					expect(
						document
							.querySelector('.tabPane:nth-child(2)')
							.className.includes('tabPaneActive'),
					).toEqual(true);
					expect(
						document
							.querySelector('.tabPane:nth-child(3)')
							.className.includes('tabPaneActive'),
					).toEqual(false);
					done();
				});
			});
		});

		it('should only display the third tab pane when tab 3 is clicked', done => {
			act(() => {
				ReactDOM.render(
					<Tabs>
						{tabData.map(tabData => (
							<Tab key={tabData} title={tabData.title}>
								{tabData.content}
							</Tab>
						))}
					</Tabs>,
					container,
				);
			});

			act(() => {
				document
					.querySelector('.tabs button:nth-child(3)')
					.dispatchEvent(evt);

				setTimeout(() => {
					expect(
						document
							.querySelector('.tabPane:nth-child(1)')
							.className.includes('tabPaneActive'),
					).toEqual(false);
					expect(
						document
							.querySelector('.tabPane:nth-child(2)')
							.className.includes('tabPaneActive'),
					).toEqual(false);
					expect(
						document
							.querySelector('.tabPane:nth-child(3)')
							.className.includes('tabPaneActive'),
					).toEqual(true);
					done();
				});
			});
		});

		it('should call onChange callback with correct active tab index', done => {
			const spyedCallback = jest.fn();

			act(() => {
				ReactDOM.render(
					<Tabs onChange={spyedCallback}>
						{tabData.map(tabData => (
							<Tab key={tabData} title={tabData.title}>
								{tabData.content}
							</Tab>
						))}
					</Tabs>,
					container,
				);
			});

			act(() => {
				document
					.querySelector('.tabs button:nth-child(2)')
					.dispatchEvent(evt);

				setTimeout(() => {
					expect(spyedCallback).toHaveBeenCalledWith(1);
					done();
				});
			});
		});
	});
});
