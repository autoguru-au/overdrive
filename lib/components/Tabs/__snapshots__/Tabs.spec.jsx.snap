// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tabs /> should allow id overriding 1`] = `
<div
  class="base"
>
  <div
    class="base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base none_mobile_base none_mobile_base 1_mobile_base none_mobile_base tabsList_root"
    role="tablist"
  >
    <button
      aria-controls="tab-1-title-0"
      aria-selected="true"
      class="base appearance cursorPointer button display_block navItem_default_base navItem_active_base"
      role="tab"
    >
      <span
        class="tabItem"
      >
        tab 1 title
      </span>
    </button>
    <button
      aria-controls="tab-2-title-1"
      aria-selected="false"
      class="base appearance cursorPointer button display_block navItem_default_base"
      role="tab"
      tabindex="-1"
    >
      <span
        class="tabItem"
      >
        tab 2 title
      </span>
    </button>
    <button
      aria-controls="tab-3-title-2"
      aria-selected="false"
      class="base appearance cursorPointer button display_block navItem_default_base"
      role="tab"
      tabindex="-1"
    >
      <span
        class="tabItem"
      >
        tab 3 title
      </span>
    </button>
  </div>
  <div
    class="base 6_mobile_base 6_mobile_base tabPanes"
  >
    <div
      class="base tabPane"
      id="tab-1-title-0"
      role="tabpanel"
      tabindex="0"
    >
      tab 1 content
    </div>
  </div>
</div>
`;

exports[`<Tabs /> should allow rendering indications 1`] = `
<div
  class="base"
>
  <div
    class="base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base none_mobile_base none_mobile_base 1_mobile_base none_mobile_base tabsList_root"
    role="tablist"
  >
    <button
      aria-controls="10-0-tab"
      aria-selected="true"
      class="base appearance cursorPointer button display_block navItem_default_base navItem_active_base"
      role="tab"
    >
      <span
        class="tabItem"
      >
        tab 1 title
      </span>
      <span
        class="navItemIndication_default_base navItemIndication_active_base"
      >
        5
      </span>
    </button>
    <button
      aria-controls="10-1-tab"
      aria-selected="false"
      class="base appearance cursorPointer button display_block navItem_default_base"
      role="tab"
      tabindex="-1"
    >
      <span
        class="tabItem"
      >
        tab 2 title
      </span>
      <span
        class="navItemIndication_default_base"
      >
        5
      </span>
    </button>
    <button
      aria-controls="10-2-tab"
      aria-selected="false"
      class="base appearance cursorPointer button display_block navItem_default_base"
      role="tab"
      tabindex="-1"
    >
      <span
        class="tabItem"
      >
        tab 3 title
      </span>
      <span
        class="navItemIndication_default_base"
      >
        5
      </span>
    </button>
  </div>
  <div
    class="base 6_mobile_base 6_mobile_base tabPanes"
  >
    <div
      class="base tabPane"
      id="10-0-tab"
      role="tabpanel"
      tabindex="0"
    >
      tab 1 content
    </div>
  </div>
</div>
`;

exports[`<Tabs /> should match snapshot (high level) 1`] = `
<div
  class="base"
>
  <div
    class="base border_style border_colour_top_gray_base border_colour_right_gray_base border_colour_bottom_gray_base border_colour_left_gray_base none_mobile_base none_mobile_base 1_mobile_base none_mobile_base tabsList_root"
    role="tablist"
  >
    <button
      aria-controls="1-0-tab"
      aria-selected="true"
      class="base appearance cursorPointer button display_block navItem_default_base navItem_active_base"
      role="tab"
    >
      <span
        class="tabItem"
      >
        tab 1 title
      </span>
    </button>
    <button
      aria-controls="1-1-tab"
      aria-selected="false"
      class="base appearance cursorPointer button display_block navItem_default_base"
      role="tab"
      tabindex="-1"
    >
      <span
        class="tabItem"
      >
        tab 2 title
      </span>
    </button>
    <button
      aria-controls="1-2-tab"
      aria-selected="false"
      class="base appearance cursorPointer button display_block navItem_default_base"
      role="tab"
      tabindex="-1"
    >
      <span
        class="tabItem"
      >
        tab 3 title
      </span>
    </button>
  </div>
  <div
    class="base 6_mobile_base 6_mobile_base tabPanes"
  >
    <div
      class="base tabPane"
      id="1-0-tab"
      role="tabpanel"
      tabindex="0"
    >
      tab 1 content
    </div>
  </div>
</div>
`;
